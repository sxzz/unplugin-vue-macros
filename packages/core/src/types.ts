export declare function defineModel<T>(): T
