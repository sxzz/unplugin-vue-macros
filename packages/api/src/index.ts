export { MagicString } from '@vue-macros/common'

// WIP
// export * from './vue'
export * from './ts'
export * from './utils'
