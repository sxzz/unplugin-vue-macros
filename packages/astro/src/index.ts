import { type AstroIntegration } from 'astro'
import { type Options } from 'unplugin-vue-macros'
import VitePlugin from 'unplugin-vue-macros/vite'

const name = 'unplugin-vue-macros'

export default function VueMacrosAstroIntegration(
  userOptions: Options = {}
): AstroIntegration {
  return {
    name,
    hooks: {
      'astro:config:setup': ({ updateConfig }) => {
        updateConfig({
          vitePlugin: [VitePlugin(userOptions)],
        })
      },
    },
  }
}
