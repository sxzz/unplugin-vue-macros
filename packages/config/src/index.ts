export * from './config-worker'
export * from './config'
export * from './define'
export * from './options'
