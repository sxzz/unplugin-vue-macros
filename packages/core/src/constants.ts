export const DEFINE_MODEL = 'defineModel'
