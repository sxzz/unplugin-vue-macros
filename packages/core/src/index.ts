import { compileScript } from '@vue/compiler-sfc'
import MagicString from 'magic-string'
import { DEFINE_MODEL } from './constants'
import { isCallOf, parseSFC } from './utils'
import type {
  Node,
  Statement,
  TSInterfaceBody,
  TSTypeLiteral,
} from '@babel/types'

export const transform = (code: string, filename: string) => {
  let hasDefineModelCall = false
  let propsTypeDecl: TSInterfaceBody | TSTypeLiteral | undefined

  function processDefineModel(node: Node) {
    if (!isCallOf(node, DEFINE_MODEL)) {
      return false
    }

    if (hasDefineModelCall) {
      throw new SyntaxError(`duplicate ${DEFINE_MODEL}() call`)
    }
    hasDefineModelCall = true

    const propsTypeDeclRaw = node.typeParameters?.params[0]
    if (!propsTypeDeclRaw) {
      throw new SyntaxError('Error: 1')
    }
    propsTypeDecl = resolveQualifiedType(
      propsTypeDeclRaw,
      (node) => node.type === 'TSTypeLiteral'
    ) as TSTypeLiteral | TSInterfaceBody | undefined

    if (!propsTypeDecl) {
      throw new Error(
        `type argument passed to ${DEFINE_MODEL}() must be a literal type, ` +
          `or a reference to an interface or literal type.`
      )
    }
  }

  function resolveQualifiedType(
    node: Node,
    qualifier: (node: Node) => boolean
  ) {
    if (qualifier(node)) {
      return node
    }
    if (
      node.type === 'TSTypeReference' &&
      node.typeName.type === 'Identifier'
    ) {
      const refName = node.typeName.name
      const isQualifiedType = (node: Node): Node | undefined => {
        if (
          node.type === 'TSInterfaceDeclaration' &&
          node.id.name === refName
        ) {
          return node.body
        } else if (
          node.type === 'TSTypeAliasDeclaration' &&
          node.id.name === refName &&
          qualifier(node.typeAnnotation)
        ) {
          return node.typeAnnotation
        } else if (node.type === 'ExportNamedDeclaration' && node.declaration) {
          return isQualifiedType(node.declaration)
        }
      }
      const body = sfc.scriptSetup!.scriptSetupAst!
      for (const node of body) {
        const qualified = isQualifiedType(node)
        if (qualified) {
          return qualified
        }
      }
    }
  }

  if (!code.includes(DEFINE_MODEL)) return

  const sfc = parseSFC(code, filename)
  if (!sfc.scriptSetup) return

  if (!sfc.scriptSetup.scriptSetupAst) {
    sfc.scriptSetup = compileScript(sfc, {
      id: filename,
    })
  }
  const { scriptSetup } = sfc
  const startOffset = scriptSetup.loc.start.offset
  // const endOffset = scriptSetup.loc.end.offset

  const s = new MagicString(code)

  for (const node of scriptSetup.scriptSetupAst as Statement[]) {
    if (node.type === 'VariableDeclaration' && !node.declare) {
      const total = node.declarations.length
      let left = total

      for (let i = 0; i < total; i++) {
        const decl = node.declarations[i]
        if (decl.init) {
          processDefineModel(decl.init)

          if (hasDefineModelCall) {
            if (left === 1) {
              s.remove(node.start! + startOffset, node.end! + startOffset)
            } else {
              let start = decl.start! + startOffset
              let end = decl.end! + startOffset
              if (i < total - 1) {
                // not the last one, locate the start of the next
                end = node.declarations[i + 1].start! + startOffset
              } else {
                // last one, locate the end of the prev
                start = node.declarations[i - 1].end! + startOffset
              }
              s.remove(start, end)
              left--
            }
          }
        }
      }
    }
  }

  if (!propsTypeDecl) return

  if (propsTypeDecl.type !== 'TSTypeLiteral') {
    throw new SyntaxError(
      `type argument passed to ${DEFINE_MODEL}() must be a literal type, or a reference to an interface or literal type.`
    )
  }

  const map: Record<string, string> = {}
  for (const member of propsTypeDecl.members) {
    if (
      member.type !== 'TSPropertySignature' ||
      member.key.type !== 'Identifier' ||
      !member.typeAnnotation?.typeAnnotation
    ) {
      throw new SyntaxError(
        `type argument passed to ${DEFINE_MODEL}() must be a literal type, or a reference to an interface or literal type.`
      )
    }

    const value = scriptSetup.loc.source.slice(
      member.typeAnnotation.typeAnnotation.start!,
      member.typeAnnotation.typeAnnotation.end!
    )

    map[member.key.name] = value
  }

  const str = `
  const props = defineProps<{
    ${Object.entries(map)
      .map(([key, value]) => `${key}: ${value}`)
      .join('\n')}
  }>()

  const emit = defineEmits<{
    ${Object.entries(map)
      .map(([key, value]) => `(evt: 'update:${key}', value: ${value}): void`)
      .join('\n')}
  }>()`

  s.prependLeft(scriptSetup.loc.start.offset, str)

  return {
    code: s.toString(),
    get map() {
      return s.generateMap()
    },
  }
}
