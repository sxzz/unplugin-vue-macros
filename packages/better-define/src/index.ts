import process from 'node:process'
import { RollupResolve, setResolveTSFileIdImpl } from '@vue-macros/api'
import {
  createFilter,
  detectVueVersion,
  FilterFileType,
  getFilterPattern,
  type BaseOptions,
  type MarkRequired,
} from '@vue-macros/common'
import {
  createUnplugin,
  type UnpluginContextMeta,
  type UnpluginInstance,
} from 'unplugin'
import { generatePluginName } from '#macros' with { type: 'macro' }
import { transformBetterDefine } from './core'
import type { PluginContext } from 'rollup'

export type Options = BaseOptions
export type OptionsResolved = MarkRequired<
  Options,
  'include' | 'version' | 'isProduction'
>

function resolveOptions(
  options: Options,
  framework: UnpluginContextMeta['framework'],
): OptionsResolved {
  const version = options.version || detectVueVersion()
  const include = getFilterPattern(
    [FilterFileType.VUE_SFC_WITH_SETUP, FilterFileType.SETUP_SFC],
    framework,
  )
  return {
    include,
    isProduction: process.env.NODE_ENV === 'production',
    ...options,
    version,
  }
}

const name = generatePluginName()

const plugin: UnpluginInstance<Options | undefined, false> = createUnplugin(
  (userOptions = {}, { framework }) => {
    const options = resolveOptions(userOptions, framework)
    const filter = createFilter(options)

    const { resolve, handleHotUpdate } = RollupResolve()

    return {
      name,
      enforce: 'pre',

      buildStart() {
        if (framework === 'rollup' || framework === 'vite') {
          setResolveTSFileIdImpl(resolve(this as PluginContext))
        }
      },

      transformInclude: filter,
      async transform(code, id) {
        try {
          return await transformBetterDefine(code, id, options.isProduction)
        } catch (error: unknown) {
          this.warn(`${name} ${error}`)
          console.warn(error)
        }
      },

      vite: {
        configResolved(config) {
          options.isProduction ??= config.isProduction
        },

        handleHotUpdate,
      },
    }
  },
)
export default plugin
