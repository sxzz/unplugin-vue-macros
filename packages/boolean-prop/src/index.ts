import { getVuePluginApi, type VuePluginApi } from '@vue-macros/common'
import { generatePluginName } from '#macros' with { type: 'macro' }
import { transformBooleanProp, type Options } from './core/index'
import type { Plugin } from 'vite'

// legacy export
export * from './api'

const name = generatePluginName()

function rollup(options: Options = {}): Plugin {
  let api: VuePluginApi | null | undefined

  return {
    name,
    configResolved(config) {
      try {
        api = getVuePluginApi(config.plugins)
      } catch {}
    },
    buildStart(rollupOpts) {
      if (api === undefined)
        try {
          api = getVuePluginApi(rollupOpts.plugins)
        } catch (error: any) {
          this.warn(error)
          return
        }

      if (!api) return

      api.options.template ||= {}
      api.options.template.compilerOptions ||= {}
      api.options.template.compilerOptions.nodeTransforms ||= []

      api.options.template.compilerOptions.nodeTransforms.push(
        transformBooleanProp(options),
      )
    },
  }
}

const plugin: {
  rollup: typeof rollup
  vite: typeof rollup
} = {
  rollup,
  vite: rollup,
}
export default plugin
