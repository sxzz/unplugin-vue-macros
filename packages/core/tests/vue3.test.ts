/* eslint-disable unicorn/prefer-string-replace-all */

import { resolve } from 'path'
import { readFile } from 'fs/promises'
import { describe, expect, it } from 'vitest'
import glob from 'fast-glob'
import { transform } from '../src'
// import { ToString, getCode } from './_utils'

describe('vue 3', () => {
  describe('fixtures', async () => {
    const root = resolve(__dirname, '..')
    const files = await glob('tests/fixtures/*.{vue,js,ts}', {
      cwd: root,
      onlyFiles: true,
    })

    for (const file of files) {
      it(file.replace(/\\/g, '/'), async () => {
        const filepath = resolve(root, file)

        const { code } = transform(await readFile(filepath, 'utf-8'), filepath)
        expect(code).toMatchSnapshot()

        // const unpluginCode = await getCode(filepath, [
        //   // VueDefineOptions({}),
        //   ToString,
        // ]).catch((err) => err)
        // expect(unpluginCode).toMatchSnapshot()
      })
    }
  })
})
