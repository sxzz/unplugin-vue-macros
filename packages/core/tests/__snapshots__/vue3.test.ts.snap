// Vitest Snapshot v1

exports[`vue 3 > fixtures > tests/fixtures/basic.vue 1`] = `
"<script setup lang=\\"ts\\">
  const props = defineProps<{
    modelValue: string
visible: boolean
  }>()

  const emit = defineEmits<{
    (evt: 'update:modelValue', value: string): void
(evt: 'update:visible', value: boolean): void
  }>()


console.log(modelValue)
modelValue = 'newValue'
</script>

<template>
  <h1>123</h1>
</template>
"
`;

exports[`vue 3 > fixtures > tests/fixtures/decls.vue 1`] = `
"<script setup lang=\\"ts\\">
  const props = defineProps<{
    modelValue: string
visible: boolean
  }>()

  const emit = defineEmits<{
    (evt: 'update:modelValue', value: string): void
(evt: 'update:visible', value: boolean): void
  }>()
let a = 'b'
</script>
"
`;
